sn_node/src/node/bootstrap/join.rs

```diff
@@ -375,6 +369,51 @@ impl<'a> Joiner<'a> {
         }
     }
 
+    // The retry responses may containing different expected ages, or only having
+    // more than 1/3 of elders voted (which blocks the aggregation).
+    // So a restart with new name shall happen when: received 1/3 of elders voted
+    // And with an age of: the most expected age
+    fn insert_retry_response(
+        &mut self,
+        sender: Peer,
+        expected_age: u8,
+        elders: BTreeSet<Peer>,
+    ) -> Option<u8> {
+        if !elders.contains(&sender) {
+            error!("Sender {sender:?} of the retry-response is not part of the elders {elders:?}");
+            return None;
+        }
+        let _ = self.retry_responses_cache.insert(sender, expected_age);
+        let mut responded = 0;
+        let mut expected_ages: BTreeMap<u8, u8> = Default::default();
+        for (elder, expected_age) in self.retry_responses_cache.iter() {
+            if elders.contains(elder) {
+                responded += 1;
+                *expected_ages.entry(*expected_age).or_insert(1) += 1;
+            }
+        }
+
+        if responded > elders.len() / 3 && !self.aggregated {
+            let mut voted_expected_age = 255;
+            let mut highest_count = 0;
+            for (expected_age, count) in expected_ages {
+                // In case two ages got the same votes, we prefer the lower age
+                if highest_count < count {
+                    highest_count = count;
+                    voted_expected_age = expected_age;
+                }
+            }
```

Comment:
That comment doesn't seem to match what the code is doing? Are you relying on the btreeset ordering the elements by age?
Can we rewrite this more clearly where we show how the ties are resolved explicitly?
```suggestion
            // In case two ages got the same votes, we prefer the lower age
            let (voted_expected_age, count) = expected_ages
                .iter()
                .max_by_key(|(age, count)| (count, -(age as i32)))?;
```

Reply:
Ok, I can understand the reluctance to filter here, but any reason why you don't use the iterators and the `max_by_key()` function? Then you won't need to add a comment to explain the BTreeMap implementation detail